src/brsTypes/components/RoAssociativeArray.ts

```diff
@@ -125,7 +154,23 @@ export class RoAssociativeArray extends BrsComponent implements BrsValue, BrsIte
             returns: ValueKind.Boolean,
         },
         impl: (interpreter: Interpreter, str: BrsString) => {
-            let deleted = this.elements.delete(str.value.toLowerCase());
+            let key = this.findElementKey(str.value, this.modeCaseSensitive);
+            let deleted = key ? this.elements.delete(key) : false;
+
+            let lKey = str.value.toLowerCase();
+            if (this.keyMap.get(lKey) === key) {
+                this.keyMap.delete(lKey);
+                // if we have {"DD": 0, "dD": 0},  keyMap["dd"] is pointed to "DD",
+                // and we delete it, then we should find another value for case insensitive accessing ("dD")
```

Comment:
Thanks for calling this out!  This is an interesting case, and feels like `keyMap` should be `Map<string, Set<string>>`: a map of lowercase keys to all encountered cases of that key.  In your example with `{ "DD": 0, "dD": 1}`, `keyMap.get("dd")` would be the set `("DD", "dD")`.  When `someAA.delete("DD")` is called from BrightScript, this function removes the first element from that set (something like `let newSet = new Set([ ...keyMap.get(lKey) ].slice(1));`) in case-sensitive mode, and removes the entire entry (`this.keyMap.delete(lKey)`) in case-insensitive mode.

Lookups become pretty similar, using the element at index zero of this `Set<string>`.

Reply:
done